### DISCUSSION TITLE
What is errors.EMFILE ?

### DISCUSSION THREAD

**Discussion by abcfy2** [artilleryio/artillery#1742] [Artillery]
When I use Lambda for testing, I found some errors I can't understand:

```sh
--------------------------------------
Metrics for period to: 12:18:30(+0800) (width: 9.999s)
--------------------------------------

errors.EINVAL: ................................................................. 4
errors.EMFILE: ................................................................. 14085
errors.ETIMEDOUT: .............................................................. 297
http.codes.201: ................................................................ 141
http.codes.500: ................................................................ 99
http.request_rate: ............................................................. 1465/sec
http.requests: ................................................................. 14646
http.response_time:
  min: ......................................................................... 5
  max: ......................................................................... 155
  median: ...................................................................... 77.5
  p95: ......................................................................... 111.1
  p99: ......................................................................... 130.3
http.responses: ................................................................ 240
```

Could you please tell me what the errors means ? Like `errors.EINVAL`, `errors.EMFILE`. Is there anything errors meaning table or list ?

**Comment by abcfy2** [comment]
And why are so many `errors.EMFILE` when I run in Lambda, but when I run in local machine, I almost can't see this error.

**Comment by juanig1** [comment]
Hello! 

Errors in caps are OS ones ([man page for linux errors](https://man7.org/linux/man-pages/man3/errno.3.html)).

Seems you are hitting an OS restriction on open files ([relevant](https://repost.aws/questions/QUP9XOYcTdQoyVq24vN1Rw7A/error-connection-issue-emfile)).

Have not happened to me, but:
- Do you have any custom logic?
- Does it still happen with a simpler script?

If you share a simple script that reproduces the issue I can look into it

**Comment by hassy** [comment]
`EMFILE` on Lambda most likely means your test is opening more TCP connections than allowed. The limit for file descriptors (which includes network connections) on Lambda is 1024 per running function (https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html).

The way to work around it is to reduce the number of arrivals in your script, and scale out horizontally instead by using more workers with the `--count` flag.

**Reply by abcfy2** [reply]
> `EMFILE` on Lambda most likely means your test is opening more TCP connections than allowed. The limit for file descriptors (which includes network connections) on Lambda is 1024 per running function (https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html).
> 
> The way to work around it is to reduce the number of arrivals in your script, and scale out horizontally instead by using more workers with the `--count` flag.

Thanks, very helpful. May I ask you what the reasonable parameters for lambda?

Here is my config:

```yaml
config:
  phases:
    - duration: 60
      arrivalRate: 5
      name: Warm up
    - duration: 120
      arrivalRate: 5
      rampTo: 150
      name: Ramp up load
    - duration: 660
      arrivalRate: 250
      maxVusers: 1000
      name: Sustained load
```

**Reply by hassy** [reply]
It really depends on the duration of the VU scenarios that are running. For example, if each VU takes 60 seconds, and you set an arrival rate of 20, then you will already exceed the 1024 limit at 52 seconds. The math is that after 52 seconds * 20 arrivals/second = 1040 VUs will be created and still running (since they take 60 seconds to complete), and each VU uses up one TCP connection.
